node/subsystem-util/src/memvisor.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Polkadot Memvisor helper for tracing memory usage across subsystems.
+
+#![forbid(unused_imports)]
+
+use metrics::Metrics;
+use std::{default::Default, sync::Arc};
+
+use crate::{
+	metrics,
+	metrics::{
+		prometheus,
+		prometheus::{CounterVec, Opts, PrometheusError, Registry, U64},
+	},
+};
+
+#[derive(Default, Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemVisorMetrics(Option<MemVisorMetricsInner>);
+
+#[derive(Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemVisorMetricsInner {
+	/// Counters for memory (de)allocation.
+	/// Labels:
+	/// - `span_name`
+	/// - `object` ... the obejct we are tracking memory for.
+	mem_alloc: CounterVec<U64>,
+	mem_free: CounterVec<U64>,
+}
+
+impl Metrics for MemVisorMetrics {
+	fn try_register(registry: &Registry) -> Result<Self, PrometheusError> {
+		let metrics = MemVisorMetricsInner {
+			mem_alloc: prometheus::register(
+				CounterVec::new(
+					Opts::new("memvisor_alloc", "Memory footprint increase for a span."),
+					&["span_name", "object"],
+				)?,
+				registry,
+			)?,
+			mem_free: prometheus::register(
+				CounterVec::new(
+					Opts::new("memvisor_free", "Memory footprint decrease for a span."),
+					&["span_name", "object"],
+				)?,
+				registry,
+			)?,
+		};
+		Ok(MemVisorMetrics(Some(metrics)))
+	}
+}
+
+#[derive(Clone)]
+#[cfg_attr(test, derive(Debug))]
+pub struct MemSpan {
+	/// The name of the span.
+	name: &'static str,
+	/// The metric value we are tracking.
+	value: u64,
+	/// Reference to MemVisorMetrics to enable metric reporting
+	/// from inside spans.
+	visor_metrics: Arc<MemVisorMetrics>,
+}
+
+#[derive(PartialEq)]
+#[cfg_attr(test, derive(Debug))]
+enum MemSpanEvent {
+	None,
+	Alloc(u64),
+	Free(u64),
```

Comment:
The naming is slightly confusing, we are not really having individual allocations or frees, we only have accumulated alloc + frees with a net positive or negative.

Reply:
It's a bit fuzzy, I agree. Doesn't sample individual (de)allocation(s) like in malloc/free, but instead it samples an arbitrary number of mallocs/frees to determine if the code inside the span has either increased or decreased the memory footprint of specific objects.